keps/sig-storage/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,681 @@
+to delete the Pod. Similar mechanism is leveraged by this proposal to automatically 
+delete the PVCs created by the StatefulSet controller.
+
+### Changes required
+
+The following changes are required:
+
+1. Add `VolumeReclaimPolicy` entry into StatefulSet spec inorder to make this feature an opt-in.
+2. Provide the following VolumeReclaimPolicies:
+   * Retain - this is the default policy and is considered in cases where no policy is specified. 
+       This would be the existing behaviour - when a StatefulSet is deleted, no action is taken with
+       respect to the PVCs created by the StatefulSet.
+   * RemoveOnPodDeletion - Whenever a pod is deleted, the corresponding PVC is deleted as well. 
+       When a scale down occurs for the StatefulSet, the corresponding PVC is deleted. A scale up 
+       following a scale down, will wait till old PVC for the removed Pod is deleted and ensure 
+       that the PVC used is a freshly created one.
+   * RemoveOnStatefulSetDeletion - PVCs corresponding to the StatefulSet are deleted when StatefulSet
+       themselves get deleted.
+3. Add `patch` to the statefulset controller cluster role.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+User environment is such at the content of the PVCs which are created automatically during StatefulSet 
+creation need not be retained after the StatefulSet is deleted. User also requires that the scale 
+up/down occurs in a fast manner, and leverages any previously existing auto created PVCs within the 
+life time of the StatefulSet. A provision needs to be provided for the user to auto-delete the PVCs 
+once the StatefulSet is deleted. 
+
+User would set the VolumeReclaimPolicy as 'RemoveOnStatefulSetDelete' which would ensure that 
+the PVCs created automatically during the StatefulSet activation is removed once the StatefulSet 
+is deleted.
+
+#### Story 2
+User is cost conscious but at the same time can sustain slower scale up(after a scale down) speeds. Needs 
+a provision where the PVC created for a pod(which is part of the StatefulSet) is removed when the Pod 
+is deleted as part of a scale down. Since the subsequent scale up needs to create fresh PVCs, it will 
+be slower than scale ups relying on existing PVCs(from earlier scale ups). 
+
+User would set the VolumeReclaimPolicy as 'RemoveOnPodDeletion' ensuring PVCs are deleted when corresponding
```

Comment:
can you confirm, you dont need RemoveOnPodDeletion, but just RemoveOnScaleDown ?

Reply:
if someone deletes the pod manually `kubectl delete pod` , do you expect the PVC to be deleted ?

Reply:
I think that makes sense. If a pod is manually deleted, the SS will recreate it so there won't be a problem of a PVC to clean up. Only on scaledown (or SS deletion) are there PVCs that currently need to be manually removed.

The two options RemoveOnScaleDown and RemoveOnStatefulSetDeletion allow the operator to distinguish between the case where scaledowns and scaleups will be frequent (and PVCs should only be deleted at the end, when the whole SS is deleted) and the case when they are infrequent and it's more important to conserve PV resources than have rapid scaleup.

At least that's where I've been coming from with this.

Reply:
Would someone want both behaviors?

Reply:
Good point. I think I implicitly assumed deletion of the stateful set is equivalent to scaling down to zero, so that RemoveOnScaleDown contains RemoveOnStatefulSetDeletion.

Probably I'm being imprecise and we should call that out explicitly.

Reply:
Looks like manual deletion needs more consideration. When a Pod is deleted, the PVC still has owner references to the deleted Pod. The StatefulSet controller will bring up the Pod again, but the reconcile logic would exit since the Pod owner is different. Eventually the GC would delete the older PVC and the next reconcile will construct new PVC and add reference to the newly created Pod. This might not be the desired behavior as we are currently talking about only deletions during scale down not during the Pod deletion.

Reply:
Good point.

I think the behavior we want is that if the SS controller is deleting a pod it knows if it's scaling down and can delete the PVC appropriately.

Otherwise the SS controller will bring the pod back up (eg if it fails) and in that case we want it to reuse the PVC even on RemoveOnScaleDown.

Will it be enough for the SS controller to only put the ownerref when it is deleting a pod for scaledown? Precisely, when the SS controller wants to remove a pod for scaledown, it first puts an ownerref on the PVC and then deletes the pod (rather than putting on the ownerref at PVC creation as we currently have it).

I don't know if this means an extra reconcile loop, where first we confirm the ownerref has been placed and only then we delete the pod?

Reply:
I learned a few things from discussions and thanks for this KEP.

> Will it be enough for the SS controller to only put the ownerref when it is deleting a pod for scaledown? Precisely, when the SS controller wants to remove a pod for scaledown, it first puts an ownerref on the PVC and then deletes the pod (rather than putting on the ownerref at PVC creation as we currently have it).

Good one but when we are patching with ownerref I think we have to set BlockOwnerRef as false. If we follow this approach there will be cases where PVC remains even after sts is scaledown. 
Ex: User deleted the sts pod and within no time(before sts controller reconciles for sts) if the user also performs scale down the sts PVC will remain the same(ownerref is not yet set). Is this case will be marked as known limitations? Correct me if my assumption is wrong in the above case.

Reply:
@mittachaitu Good point, I'd have to look at the statefulset controller but you're probably right.

Reply:
@kk-src and I looked through the current [SS controller code](https://github.com/kubernetes/kubernetes/blob/64f8bf24a6d74de9a42b10427ceb0e0ced304fea/pkg/controller/statefulset/stateful_set_control.go#L315) and came to the conclusion that if a manual deletion races with the scaledown, the controller will not do the scaledown and instead wait for all the pods to come back. So I don't think we need to change the proposed logic.

If the race happens in the other direction (user deleted after we start scaling down), it's okay as we're only setting the owner refs for PVCs with pods that are going to be deleted anyway.

At least that's how I reasoned it through, LMK if I missed something.